addons/certificate/models/certificate.py

```diff
@@ -0,0 +1,281 @@
+        attachment=False,
+        store=True,
+        compute='_compute_pem_certificate'
+    )
+    serial_number = fields.Char(
+        string='Serial number',
+        help='The serial number to add to electronic documents',
+        compute='_compute_pem_certificate',
+        store=True,
+    )
+    date_start = fields.Datetime(
+        string='Available date',
+        help='The date on which the certificate starts to be valid (UTC)',
+        compute='_compute_pem_certificate',
+        store=True,
+    )
+    date_end = fields.Datetime(
+        string='Expiration date',
+        help='The date on which the certificate expires (UTC)',
+        compute='_compute_pem_certificate',
+        store=True,
+    )
+    is_valid = fields.Boolean(string='Valid', compute='_compute_is_valid', search='_search_is_valid')
+    active = fields.Boolean(name='Active', help='Set active to false to archive the certificate.', default=True)
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        string='Company',
+        required=True,
+        default=lambda self: self.env.company,
+        ondelete='cascade',
+    )
+    country_code = fields.Char(related='company_id.country_id.code', depends=['company_id'])
+
+    # Technical field used to display a certificate loading error
+    loading_error = fields.Text(string='Loading error', compute="_compute_pem_certificate", store=True)
+
+    @api.depends('content', 'pkcs12_password')
+    def _compute_pem_certificate(self):
+        for certificate in self:
+            if not certificate.content:
+                certificate.pem_certificate = None
+                certificate.date_start = None
+                certificate.date_end = None
+                certificate.serial_number = None
+                certificate.private_key = None
+                certificate.public_key = None
+                certificate.loading_error = ""
+                certificate.scope = certificate.scope or certificate.env.context.get('scope', 'general')
+            else:
+                content = base64.b64decode(certificate.content)
+                cert = key = None
+
+                # Try to load the certificate in different format starting with DER then PKCS12 and
+                # finally PEM. If none succeeded, we report an error.
+                try:
+                    cert = x509.load_der_x509_certificate(content)
+                    certificate.content_format = 'der'
+                except ValueError:
+                    pass
+                if not cert:
+                    try:
+                        pkcs12_password = certificate.pkcs12_password.encode('utf-8') if certificate.pkcs12_password else None
+                        key, cert, _additional_certs = pkcs12.load_key_and_certificates(content, pkcs12_password)
+                        certificate.content_format = 'pkcs12'
+                    except ValueError:
+                        pass
```

Comment:
pass is really doing nothing?  Should not we put the later code instead and avoid all the if not cert:?  (it would have some nested try..except

Reply:
I believe nested try except to be highly unreadable for the extremely low improvement gain. Furthermore, it's easier to add or remove certificate formats that way (not that it's gonna happen but still).

So if that's okay with you, I prefer to keep it that way :+1: